rpcs3/Emu/Cell/lv2/sys_spu.cpp

```diff
@@ -47,38 +62,45 @@ void sys_spu_image::load(const fs::file& stream)
 		LOG_NOTICE(SPU, "** Section: sh_type=0x%x, addr=0x%llx, size=0x%llx, flags=0x%x", shdr.sh_type, shdr.sh_addr, shdr.sh_size, shdr.sh_flags);
 	}
 
-	for (const auto& prog : obj.progs)
+	for (auto it_prog = obj.progs.begin(); it_prog < obj.progs.end(); it_prog++)
 	{
+		auto& prog = *it_prog;
+
 		sha1_update(&ctx, reinterpret_cast<const u8*>(&prog), sizeof(spu_exec_object::phdr_t));
 		sha1_update(&ctx, reinterpret_cast<const u8*>(prog.bin.data()), prog.bin.size());
 
-		LOG_NOTICE(SPU, "** Segment: p_type=0x%x, p_vaddr=0x%llx, p_filesz=0x%llx, p_memsz=0x%llx, flags=0x%x", prog.p_type, prog.p_vaddr, prog.p_filesz, prog.p_memsz, prog.p_flags);
-
+		LOG_NOTICE(SPU, "** Segment: p_type=0x%x, p_offset=0x%llx, p_vaddr=0x%llx, p_filesz=0x%llx, p_memsz=0x%llx, flags=0x%x", prog.p_type, prog.p_offset, prog.p_vaddr, prog.p_filesz, prog.p_memsz, prog.p_flags);
 		if (prog.p_type == SYS_SPU_SEGMENT_TYPE_COPY)
 		{
-			auto& seg = segs[nsegs++];
-			seg.type  = prog.p_type;
-			seg.ls    = prog.p_vaddr;
-			seg.addr  = addr + prog.p_vaddr;
-			seg.size  = std::min(prog.p_filesz, prog.p_memsz);
-			std::memcpy(vm::base(seg.addr), prog.bin.data(), seg.size);
+			if (prog.p_filesz != 0)
+			{
+				auto& seg = segs[nsegs++];
+
+				seg.type = SYS_SPU_SEGMENT_TYPE_COPY;
+				seg.addr = data_addr + prog.p_offset;
+				seg.ls = prog.p_vaddr;
+				seg.size = prog.p_filesz;
+			}
 
-			if (prog.p_memsz > prog.p_filesz)
+			// There's a mismatch between here and the seg counting. It seems that it exists in the real function, too.
```

Comment:
What does it mean? There shouldn't be any mismatch except two segments out of one due to zero fill.